lib/assessment.sql

```diff
@@ -56,3 +56,65 @@ SET
     grading_needed = FALSE
 WHERE
     ai.id = $assessment_instance_id;
+
+-- BLOCK select_assessments_for_statistics_update
+SELECT
+    a.id AS assessment_id
+FROM
+    assessments AS a
+WHERE
+    a.course_instance_id = $course_instance_id
+    AND EXISTS (
+        SELECT 1
+        FROM assessment_instances AS ai
+        WHERE
+            ai.assessment_id = a.id
+            AND ai.modified_at > a.statistics_last_updated_at - interval '1 minute'
+    );
+
+-- BLOCK select_assessment_lock
+SELECT 1
+FROM assessments AS a
+WHERE a.id = $assessment_id
+FOR UPDATE;
+
+-- BLOCK select_assessment_needs_statisics_update
+SELECT
+    EXISTS (
+        SELECT 1
+        FROM assessment_instances AS ai
+        WHERE
+            ai.assessment_id = a.id
+            AND ai.modified_at > a.statistics_last_updated_at - interval '1 minute'
+    ) AS needs_statistics_update
+FROM assessments AS a
+WHERE a.id = $assessment_id;
+
+-- BLOCK update_assessment_statisics
+UPDATE assessments AS a
+SET
+    statistics_last_updated_at = now(),
+    score_stat_number = score_stats.number,
+    score_stat_min = score_stats.min,
+    score_stat_max = score_stats.max,
+    score_stat_mean = score_stats.mean,
+    score_stat_std = score_stats.std,
+    score_stat_median = score_stats.median,
+    score_stat_n_zero = score_stats.n_zero,
+    score_stat_n_hundred = score_stats.n_hundred,
+    score_stat_n_zero_perc = score_stats.n_zero_perc,
+    score_stat_n_hundred_perc = score_stats.n_hundred_perc,
+    score_stat_score_hist = score_stats.score_hist,
+    duration_stat_min = duration_stats.min,
+    duration_stat_max = duration_stats.max,
+    duration_stat_mean = duration_stats.mean,
+    duration_stat_median = duration_stats.median,
+    duration_stat_thresholds = duration_stats.thresholds,
+    duration_stat_threshold_seconds = duration_stats.threshold_seconds,
+    duration_stat_threshold_labels = duration_stats.threshold_labels,
+    duration_stat_hist = duration_stats.hist
+FROM
+    assessments_score_stats($assessment_id) AS score_stats,
+    assessments_duration_stats($assessment_id) AS duration_stats
+WHERE
+    a.id = $assessment_id;
```

Comment:
Given this is the only place that these sprocs are used, wouldn't it be better to move this update to the sprocs themselves? Maybe even the check if the update is needed, to reduce the back-and-forth between Postgres and JS?

Reply:
Over the long term I'm trying to slowly move away from sprocs and towards regular queries, with typescript to check the returned data types. In this case I would ideally remove the two `_stats()` sprocs in favor of queries in `assessment.sql`. I didn't do this because I wanted to minimize the risk of this PR breaking the statistics calculation, so I left the sprocs untouched.